### ISSUE TITLE
Memory leak with compress + encrypt backend on Google Drive

### ISSUE THREAD

**Issue by mrtolkien** [rclone/rclone#6013]
#### What is the problem you are having with rclone?
`rclone` memory usage goes up linearly to the point where it crashes my machine. Tried it 5 times already, always the same outcome.

#### What is your rclone version (output from `rclone version`)
```shell
❯ rclone --version
rclone v1.58.0-beta.6000.5650c1aa3.fix-5994-compress-crash
- os/version: debian 10.11 (64 bit)
- os/kernel: 5.4.166-1-pve (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.17.6
- go/linking: static
- go/tags: none
```

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)
Debian 10 amd64

#### Which cloud storage system are you using? (e.g. Google Drive)
Google Drive

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)
`rclone copy /slowpool gdrive_tachikoma_compress:slowpool -P --skip-links`

The remotes are configured as follow:
```
[gdrive_tachikoma]
type = drive
client_id = xxxx
client_secret = xxxx
scope = drive
token = xxxx
team_drive = 

[gdrive_tachikoma_crypt]
type = crypt
remote = gdrive_tachikoma:NAS/BACKUP
password = xxx
password2 = xxx

[gdrive_tachikoma_compress]
type = compress
remote = gdrive_tachikoma_crypt:
```

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)
There are no particular logs from `rclone` as everything goes fine until the OS sends a `SIGKILL` once it reaches 100% RAM usage.

I will re-run the command with `-vv` and update here if anything of notice appears.

#### `atop` data

`VGROW` represent "The amount of virtual memory that the process has grown during the last interval". As you can see, it's roughly 1Gb/10m, which is indicative of a memory leak.

16:55
![image](https://user-images.githubusercontent.com/8489650/155836522-b8fe6895-6feb-4ef8-9622-c187c635d405.png)
17:05
![image](https://user-images.githubusercontent.com/8489650/155836531-0fa9eb03-d5bd-4aa0-adf7-67a82b6db90d.png)
17:15
![image](https://user-images.githubusercontent.com/8489650/155836576-6967a402-af14-4b83-bbcf-2a75935efe16.png)
17:25
![image](https://user-images.githubusercontent.com/8489650/155836584-b24d7ce1-f02a-4662-b1c6-41ada8ff9c05.png)
17:35
![image](https://user-images.githubusercontent.com/8489650/155836590-b7b36754-0629-4906-b595-c183e6f36258.png)

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by animosity22** [comment]
You can check it out with:

https://rclone.org/rc/#debugging-rclone-with-pprof

And a debug log would be great.

**Comment by mrtolkien** [comment]
I saved the log, and it's only the standard "checking file" ad vitam eternam until there's nothing and the process gets killed.

I'm not sure how to exactly show the issue as it will only be apparent on *huge* runs that are long enough to saturate a machine's RAM. I also don't have the luxury to crash my home server every 4 hours so it's hard for me to run other tests to failure.

**Comment by mrtolkien** [comment]
> Need the memory item I linked above if you'd like to find a cause.

Running it with `watch -n 600 go tool pprof -text http://localhost:5572/debug/pprof/heap` right now, will share the logs folder once it's a bit fuller. Also paired it with `export GOGC=20`.

<details>
  <summary>Expand to see logs after 10 minutes</summary>

```
File: rclone
Type: inuse_space
Time: Feb 27, 2022 at 10:38am (JST)
Showing nodes accounting for 234.62MB, 95.41% of 245.89MB total
Dropped 96 nodes (cum <= 1.23MB)
      flat  flat%   sum%        cum   cum%
  178.58MB 72.62% 72.62%   178.58MB 72.62%  github.com/buengese/sgzip.NewWriterLevel.func1
      40MB 16.27% 88.89%       40MB 16.27%  github.com/rclone/rclone/backend/compress.(*Fs).rcat
       8MB  3.25% 92.15%        8MB  3.25%  path.Join
       3MB  1.22% 93.37%        3MB  1.22%  encoding/json.(*decodeState).literalStore
       2MB  0.81% 94.18%        2MB  0.81%  fmt.Sprintf
    1.50MB  0.61% 94.79%     1.50MB  0.61%  github.com/rclone/rclone/backend/compress.(*Fs).newObjectSizeAndNameOnly (inline)
       1MB  0.41% 95.20%        3MB  1.22%  github.com/rclone/rclone/backend/drive.(*Fs).newRegularObject
    0.54MB  0.22% 95.41%     9.04MB  3.68%  github.com/rclone/rclone/backend/drive.(*Fs).List.func1
         0     0% 95.41%   174.96MB 71.15%  bytes.(*Reader).WriteTo
         0     0% 95.41%        3MB  1.22%  encoding/json.(*Decoder).Decode
         0     0% 95.41%        3MB  1.22%  encoding/json.(*decodeState).array
         0     0% 95.41%        3MB  1.22%  encoding/json.(*decodeState).object
         0     0% 95.41%        3MB  1.22%  encoding/json.(*decodeState).unmarshal
         0     0% 95.41%        3MB  1.22%  encoding/json.(*decodeState).value
         0     0% 95.41%   176.16MB 71.64%  github.com/buengese/sgzip.(*Writer).Write
         0     0% 95.41%     1.85MB  0.75%  github.com/buengese/sgzip.(*Writer).compressBlock
         0     0% 95.41%    15.54MB  6.32%  github.com/rclone/rclone/backend/compress.(*Fs).List
         0     0% 95.41%   214.96MB 87.42%  github.com/rclone/rclone/backend/compress.(*Fs).Put
         0     0% 95.41%     2.50MB  1.02%  github.com/rclone/rclone/backend/compress.(*Fs).addData
         0     0% 95.41%     2.50MB  1.02%  github.com/rclone/rclone/backend/compress.(*Fs).processEntries
         0     0% 95.41%       40MB 16.27%  github.com/rclone/rclone/backend/compress.(*Fs).putCompress
         0     0% 95.41%     2.41MB  0.98%  github.com/rclone/rclone/backend/compress.(*Fs).putCompress.func1
         0     0% 95.41%       40MB 16.27%  github.com/rclone/rclone/backend/compress.(*Fs).putWithCustomFunctions
         0     0% 95.41%   174.96MB 71.15%  github.com/rclone/rclone/backend/compress.checkCompressAndType
         0     0% 95.41%   174.96MB 71.15%  github.com/rclone/rclone/backend/compress.isCompressible
         0     0% 95.41%    13.04MB  5.30%  github.com/rclone/rclone/backend/crypt.(*Fs).List
         0     0% 95.41%    12.04MB  4.90%  github.com/rclone/rclone/backend/drive.(*Fs).List
         0     0% 95.41%        3MB  1.22%  github.com/rclone/rclone/backend/drive.(*Fs).itemToDirEntry
         0     0% 95.41%    12.04MB  4.90%  github.com/rclone/rclone/backend/drive.(*Fs).list
         0     0% 95.41%        3MB  1.22%  github.com/rclone/rclone/backend/drive.(*Fs).list.func2
         0     0% 95.41%        3MB  1.22%  github.com/rclone/rclone/backend/drive.(*Fs).newObjectWithInfo
         0     0% 95.41%     2.50MB  1.02%  github.com/rclone/rclone/backend/local.(*Fs).List
         0     0% 95.41%     2.50MB  1.02%  github.com/rclone/rclone/backend/local.(*Fs).cleanRemote
         0     0% 95.41%        3MB  1.22%  github.com/rclone/rclone/fs.pacerInvoker
         0     0% 95.41%    19.04MB  7.74%  github.com/rclone/rclone/fs/list.DirSorted
         0     0% 95.41%    19.04MB  7.74%  github.com/rclone/rclone/fs/march.(*March).makeListDir.func1
         0     0% 95.41%     2.50MB  1.02%  github.com/rclone/rclone/fs/march.(*March).processJob.func1
         0     0% 95.41%    16.54MB  6.73%  github.com/rclone/rclone/fs/march.(*March).processJob.func2
         0     0% 95.41%   214.96MB 87.42%  github.com/rclone/rclone/fs/operations.Copy
         0     0% 95.41%   214.96MB 87.42%  github.com/rclone/rclone/fs/sync.(*syncCopyMove).pairCopyOrMove
         0     0% 95.41%        3MB  1.22%  github.com/rclone/rclone/lib/pacer.(*Pacer).Call
         0     0% 95.41%        3MB  1.22%  github.com/rclone/rclone/lib/pacer.(*Pacer).call
         0     0% 95.41%        3MB  1.22%  google.golang.org/api/drive/v3.(*FilesListCall).Do
         0     0% 95.41%        3MB  1.22%  google.golang.org/api/internal/gensupport.DecodeResponse (inline)
         0     0% 95.41%   176.16MB 71.64%  io.Copy (inline)
         0     0% 95.41%   176.16MB 71.64%  io.copyBuffer
         0     0% 95.41%     4.13MB  1.68%  runtime.doInit
         0     0% 95.41%     4.63MB  1.88%  runtime.main
         0     0% 95.41%   179.22MB 72.88%  sync.(*Pool).Get
```
</details>

<details>
<summary>Expand to see the atop data just before my latest crash</summary>

I ran `rclone` against last night. It got to 15.4Gb RAM usage and crashed the machine again:
<img width="1397" alt="image" src="https://user-images.githubusercontent.com/8489650/155864869-01daa29c-8121-4455-88f6-2b1ca8da2ea1.png">

</details>

> Can you try setting GOGC as mentioned here and see if that improves things

Trying it as we speak!

> Try reducing --transfers and --checkers to use less memory.

Using default values on a 16Gb machine, I feel like it should not be an issue.

> In a sync the number of items in the largest directory usually controls the max memory usage. How many do you think that is?

Normally nothing crazy. I'm dumping millions of files but each directory should be of a reasonable size.

**Comment by mrtolkien** [comment]
Here is the output of `pprof` after running for one hour:

```
File: rclone
Type: inuse_space
Time: Feb 27, 2022 at 11:58am (JST)
Showing nodes accounting for 4.60GB, 98.73% of 4.65GB total
Dropped 134 nodes (cum <= 0.02GB)
      flat  flat%   sum%        cum   cum%
    4.54GB 97.47% 97.47%     4.54GB 97.47%  github.com/buengese/sgzip.NewWriterLevel.func1
    0.06GB  1.26% 98.73%     0.06GB  1.28%  github.com/rclone/rclone/backend/compress.(*Fs).rcat
         0     0% 98.73%     4.46GB 95.77%  bytes.(*Reader).WriteTo
         0     0% 98.73%     0.03GB  0.61%  github.com/buengese/sgzip.(*Writer).Close
         0     0% 98.73%     4.48GB 96.35%  github.com/buengese/sgzip.(*Writer).Write
         0     0% 98.73%     0.06GB  1.20%  github.com/buengese/sgzip.(*Writer).compressBlock
         0     0% 98.73%     0.03GB  0.61%  github.com/buengese/sgzip.(*Writer).compressCurrent
         0     0% 98.73%     4.53GB 97.26%  github.com/rclone/rclone/backend/compress.(*Fs).Put
         0     0% 98.73%     0.07GB  1.45%  github.com/rclone/rclone/backend/compress.(*Fs).putCompress
         0     0% 98.73%     0.06GB  1.19%  github.com/rclone/rclone/backend/compress.(*Fs).putCompress.func1
         0     0% 98.73%     0.07GB  1.46%  github.com/rclone/rclone/backend/compress.(*Fs).putWithCustomFunctions
         0     0% 98.73%     4.46GB 95.82%  github.com/rclone/rclone/backend/compress.checkCompressAndType
         0     0% 98.73%     4.46GB 95.82%  github.com/rclone/rclone/backend/compress.isCompressible
         0     0% 98.73%     4.53GB 97.28%  github.com/rclone/rclone/fs/operations.Copy
         0     0% 98.73%     4.53GB 97.28%  github.com/rclone/rclone/fs/sync.(*syncCopyMove).pairCopyOrMove
         0     0% 98.73%     4.48GB 96.35%  io.Copy (inline)
         0     0% 98.73%     4.48GB 96.35%  io.copyBuffer
         0     0% 98.73%     4.56GB 97.91%  sync.(*Pool).Get
```

There is a very clear culprit in:
`github.com/buengese/sgzip.NewWriterLevel.func1`

The RAM usage of this node goes up and up without ever being garbage collected.

**Comment by ncw** [comment]
> There is a very clear culprit in: `github.com/buengese/sgzip.NewWriterLevel.func1`

Ah ha! A smoking gun! Well done :-) It looks like the `isCompressible` function never closes the compressor which would explain the memory leak.

Can you try this fix?

[v1.58.0-beta.6009.f2d7985e5.fix-6013-compress-memory](https://beta.rclone.org/branch/fix-6013-compress-memory/v1.58.0-beta.6009.f2d7985e5.fix-6013-compress-memory/) on branch [fix-6013-compress-memory](https://github.com/rclone/rclone/tree/fix-6013-compress-memory) (uploaded in 15-30 mins)

**Comment by mrtolkien** [comment]
@ncw I'm not seeing a Linux amd64 build unfortunately. Do you know what that could be due to?

**Comment by ncw** [comment]
I re-ran the build as it locked up :-(

Hopefully it will be OK this time.

**Comment by ncw** [comment]
Looks like it worked this time and made binaries for you to try @mrtolkien

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.58

**Comment by mrtolkien** [comment]
I did download and run it, and it seems fixed! Thanks a lot for the quick fix.
